### ISSUE TITLE
Functions implemented on xla

### ISSUE THREAD

**Issue by KonstantinosMarko** [pytorch/xla#1509]
## ❓ Questions and Help
Hi there!
I need to train my model on TPUs. A couple of months ago i asked on why my model runs so slow, and one of the problems was that some of my functions were not implemented on xla and i had to pass my data from the TPU to CPU and the opposite. 

Some of the functions that i need are pack_padded_sequence, pad_packed_sequence.

Have they been implemented on xla yet, or how can find which functions are implemented or not implemented on xla?

Thank you very much in advance!

**Comment by dlibenzi** [comment]
Do you mind running your model using `debug_run.py` and attach the resulting tar.gz?
Just let it run for 4..5 steps and hit `CRTL^C`.

https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information

**Comment by KonstantinosMarko** [comment]
Hi @dlibenzi. Thank you for your answer!!
I attached the tar.gz file you asked for.

Thank you in advance :)

[debug_run.tar.gz](https://github.com/pytorch/xla/files/4016341/debug_run.tar.gz)

**Comment by dlibenzi** [comment]
The function that makes you exit from graph mode is `aten::masked_select`.
We currently have it in experimental mode.
If you fetch nightly wheels/dockers and use pytorch-nightly on the TPU VM, you can:
```Shell
export XLA_EXPERIMENTAL="nonzero:masked_select"
```
To see if that helps.

But your code (maybe for debugging) seems to be doing a lot of `item()` calls (or printing XLA tensor values).
Those are OK for debugging, but they can slow down a lot execution.
The way we suggest is to report (print, write to tensorboard, etc...) things every N steps:

https://github.com/pytorch/xla/blob/b9ade635e9924789d421620a9598f383221bbf30/test/test_train_mp_mnist.py#L129

You can see that your model only did 6 steps:
```
Counter: MarkStep
  Value: 6
```

But you had about 150 executions:
```
Metric: ExecuteTime
  TotalSamples: 149
  Accumulator: 43s829ms887.406us
  ValueRate: 194ms250.126us / second
  Rate: 0.675788 / second
  Percentiles: 1%=002ms708.978us; 5%=002ms919.239us; 10%=002ms45.376us; 20%=002ms224.312us; 50%=003ms671.686us; 80%=407ms961.254us; 90%=854ms240.332us; 95%=02s656ms286.751us; 99%=04s699ms120.662us
```

Usually the number of executions should track pretty closely the step count.

**Comment by KonstantinosMarko** [comment]
thank you for your answer!!

i commented every debugging metric and everything which is not needed. i also used the export you proposed. 

the execute time is the following:

Metric: ExecuteTime
  TotalSamples: 29
  Accumulator: 28s280ms238.525us
  ValueRate: 181ms833.260us / second
  Rate: 0.185436 / second
  Percentiles: 1%=003ms30.883us; 5%=003ms137.124us; 10%=004ms827.684us; 20%=114ms954.951us; 50%=477ms840.522us; 80%=02s536ms820.167us; 90%=02s813ms640.376us; 95%=04s632ms11.950us; 99%=06s271ms671.802us

But still the program runs so slow. for example:

Iter: 7, Batch: [7/5273], Batch_loss: 2.4581, Speed: 0.01 steps/s, Time: 9m41.3s

Is this because some of the functions used are not implemented in xla yet? So, the calculations travel from TPUs to CPU and backwards?

thank you in advance

**Comment by dlibenzi** [comment]
If you post again the result of `debug_run.py` we can take a look 😉

**Comment by KonstantinosMarko** [comment]
sorry for the delay. i am posting the result of debug_run.py again. how can i find out the device each calculation runs on?
[debug_run.tar.gz](https://github.com/pytorch/xla/files/4029538/debug_run.tar.gz)

**Comment by dlibenzi** [comment]
This debug-run archive looks incomplete to me.
The metrics have very little data.
The `aten::*` you seen in the graph files are OK. They all are.
The ones to be looked at, are the ones in the metrics file.

**Comment by KonstantinosMarko** [comment]
thank you for your answer. In the metrics folder the only problem i see is in L1Loss and loss.backward. Have they been implemented in xla or should i use an other loss, NLLLoss for example? Is there a problem with L1Loss and xla?
thank you!

**Comment by dlibenzi** [comment]
Guys, I know we say `0.5` is `stable`, but you really should use `nightly` (and `pytorch-nightly` for the TPU VM) 😉 
We do implement L1 loss:

https://github.com/pytorch/xla/blob/d62a57564978bdb4af56317aed454fdf0b2b62ff/torch_xla/csrc/aten_xla_type.cpp#L1424

**Comment by KonstantinosMarko** [comment]
ok i see that. but when i use 0.5, terminal informs me on using TPUs with the following way:

2020-01-08 14:40:23.340077: I    6371 torch_xla/csrc/aten_xla_type.cpp:75] PyTorch GIT revision: 19ab5381c3fa4e2205a53febf94f616143467401
2020-01-08 14:40:23.340137: I    6371 torch_xla/csrc/aten_xla_type.cpp:76] XLA GIT revision: 76f578dadf37671a0c7b558c40462816cd7e1fb5
2020-01-08 14:40:23.355784: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:217] XRT device (LOCAL) CPU:0 -> /job:tpu_worker/replica:0/task:0/device:XLA_CPU:0
2020-01-08 14:40:23.355838: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:217] XRT device (LOCAL) TPU:0 -> /job:tpu_worker/replica:0/task:0/device:TPU:0
2020-01-08 14:40:23.355847: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:217] XRT device (LOCAL) TPU:1 -> /job:tpu_worker/replica:0/task:0/device:TPU:1
2020-01-08 14:40:23.355854: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:217] XRT device (LOCAL) TPU:2 -> /job:tpu_worker/replica:0/task:0/device:TPU:2
2020-01-08 14:40:23.355863: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:217] XRT device (LOCAL) TPU:3 -> /job:tpu_worker/replica:0/task:0/device:TPU:3
2020-01-08 14:40:23.355879: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:217] XRT device (LOCAL) TPU:4 -> /job:tpu_worker/replica:0/task:0/device:TPU:4
2020-01-08 14:40:23.355901: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:217] XRT device (LOCAL) TPU:5 -> /job:tpu_worker/replica:0/task:0/device:TPU:5
2020-01-08 14:40:23.355912: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:217] XRT device (LOCAL) TPU:6 -> /job:tpu_worker/replica:0/task:0/device:TPU:6
2020-01-08 14:40:23.355926: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:217] XRT device (LOCAL) TPU:7 -> /job:tpu_worker/replica:0/task:0/device:TPU:7
2020-01-08 14:40:23.355958: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:221] Worker grpc://10.240.0.2:8470 for /job:tpu_worker/replica:0/task:0
2020-01-08 14:40:23.355971: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:225] XRT default device: TPU:0
2020-01-08 14:40:23.355989: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:1114] Configuring TPU for master worker tpu_worker:0 at grpc://10.240.0.2:8470
2020-01-08 14:40:23.372116: E    6371 tensorflow/core/framework/op_kernel.cc:1579] OpKernel ('op: "ErfinvGrad" device_type: "CPU" constraint { name: "T" allowed_values { list { type: DT_DOUBLE } } }') for unknown op: ErfinvGrad
2020-01-08 14:40:23.372195: E    6371 tensorflow/core/framework/op_kernel.cc:1579] OpKernel ('op: "ErfinvGrad" device_type: "CPU" constraint { name: "T" allowed_values { list { type: DT_FLOAT } } }') for unknown op: ErfinvGrad
2020-01-08 14:40:23.372211: E    6371 tensorflow/core/framework/op_kernel.cc:1579] OpKernel ('op: "NdtriGrad" device_type: "CPU" constraint { name: "T" allowed_values { list { type: DT_DOUBLE } } }') for unknown op: NdtriGrad
2020-01-08 14:40:23.372236: E    6371 tensorflow/core/framework/op_kernel.cc:1579] OpKernel ('op: "NdtriGrad" device_type: "CPU" constraint { name: "T" allowed_values { list { type: DT_FLOAT } } }') for unknown op: NdtriGrad
2020-01-08 14:40:27.969532: I    6371 tensorflow/compiler/xla/xla_client/xrt_computation_client.cc:1125] TPU topology: mesh_shape: 2
mesh_shape: 2
mesh_shape: 2
num_tasks: 1
num_tpu_devices_per_task: 8
device_coordinates: 0
device_coordinates: 0
device_coordinates: 0
device_coordinates: 0
device_coordinates: 0
device_coordinates: 1
device_coordinates: 0
device_coordinates: 1
device_coordinates: 0
device_coordinates: 0
device_coordinates: 1
device_coordinates: 1
device_coordinates: 1
device_coordinates: 0
device_coordinates: 0
device_coordinates: 1
device_coordinates: 0
device_coordinates: 1
device_coordinates: 1
device_coordinates: 1
device_coordinates: 0
device_coordinates: 1
device_coordinates: 1
device_coordinates: 1

Total model parameters: 9822979
Total model trainable parameters: 9822979

---> Using already extracted features...

---> Training new model from scratch...

--- Current schedule ---
| r = 7 | lr = 0.001 | max_iters = 10k | batch_size = 8 |


but when i use the torch-xla-nightly environment the only information i get is the following. there is no info on TPUs.

Total model parameters: 9822979
Total model trainable parameters: 9822979

---> Using already extracted features...

---> Training new model from scratch...

--- Current schedule ---
| r = 7 | lr = 0.001 | max_iters = 10k | batch_size = 8 |

**Comment by KonstantinosMarko** [comment]
ok. thank you very much for the info. i am attaching the complete debugging logs, on the nightly env. If you had some time to look at it and share some information or hints in order to succeed running my model, i would appreciate that. Thank you in advance :)
[debug_run.tar.gz](https://github.com/pytorch/xla/files/4035969/debug_run.tar.gz)

**Comment by dlibenzi** [comment]
Now functions looks good, and `item()` calls track the step count.
But there seem to be too many compilations.
As the number of steps being captured is low, cannot tell if it has stabilized or not.
Can you run for 15..20 steps?

**Comment by KonstantinosMarko** [comment]
i attached what you asked for (not everything because of size). i observed that after the 10th iteration, batch loss becomes NaN. The same thing happens from the 1st iteration when using batch size larger than 8. This is very weird.
[logs.zip](https://github.com/pytorch/xla/files/4040430/logs.zip)

**Comment by dlibenzi** [comment]
The `Nan` is a different issue, you may want to open a new issue for that, but mind that w/out a proper repro code (full code), it is impossible for us to debug.

For this issue, your graph keeps recompiling, and AFAICT it is due to input shapes changing across different steps.

**Comment by KonstantinosMarko** [comment]
@dlibenzi i figured out what was the problem. my data are of different length, so it happened what you wrote above. when i changed my data so every one has the same length, the training became faster. not that fast, but faster. what can i do in order to use data that have different shapes?
moreover, when i commented pack_padded and packed_pad functions, the training became even faster. So what can i do with the different shapes and the functions not supported yet in xla? thank you very much!

**Comment by dlibenzi** [comment]
I am confused now.
Which functions are not implemented?

For sequences of different length, one trick is to categorize them according to frequency, and pad them.
Like all the ones 1..16 will be padded to 16, all the ones from 17..32 will be padded to 32, ...
The smallest sequence length buckets can then use higher batch size.

About the padded functions, we will have to look into them as we also have another issue open: https://github.com/pytorch/xla/issues/1522
Unfortunately team is small and busy.

**Comment by visionscaper** [comment]
Hi @dlibenzi, I added complete working code samples for issue #1522, one using a TPU and the other a GPU. I hope this will help. Let me know if I can be of more assistance.

**Comment by dlibenzi** [comment]
> Hi @dlibenzi, I added complete working code samples for issue #1522, one using a TPU and the other a GPU. I hope this will help. Let me know if I can be of more assistance.

Thanks. We'll try to take a look ASAP.
Sorry for the delay, but small team, lot's of stuff to do 😐

**Comment by visionscaper** [comment]
No worries, I appreciate the work you do.

**Comment by KonstantinosMarko** [comment]
hi @dlibenzi 

i just found out a problem that i have. if you had any clue that could help me i would appreciate that. 

i have made some progress on the training, but i found out that when my first epoch finishes, then the program freezes. do you have any clue why this is happening?

my training loop is the following:

    def _train_step(self, batch, r):

        tracker = xm.RateTracker()

        # Prepare model and optimizer
        self.model.train()
        self.optimizer.zero_grad()

        # Prepare data
        fname, text, mel, gate, text_len = prepare_batch(batch, device=self.device)
        print(text.size())

        # Pass the data through the model and get results
        _, dec_result = self.model(text, mel, text_len, r)
        mel_out, postnet_out, _, gate_predictions = dec_result

        # Calculate losses
        mel_loss1 = self.criterion(postnet_out, mel)
        mel_loss2 = self.criterion(mel_out, mel)
        mel_loss = mel_loss1 + mel_loss2
        #print("###################################################")
        #print('mel loss: ')
        #print(mel_loss)
        #print("###################################################")
        gate_loss = self.gate_loss(gate_predictions, gate)
        #print("gate loss: ")
        #print(gate_loss)
        loss = gate_loss + mel_loss
        #print("###################################################")
        #print("Loss is:")
        #print("###################################################")
        #print(loss)
        #print("###################################################")
        loss = loss.to(device)
        #print("after loss.to(device)")
        #print(torch_xla._XLAC._get_xla_tensors_text([loss]))
        #print("before loss.backward")
        loss.backward()
        #print("after loss.backward")
        import torch_xla.debug.metrics as met
        #print(met.metrics_report())
        # Update optimizer
        #print("before clip grad norm")
        nn.utils.clip_grad_norm_(self.model.parameters(), self.clip_norm)
        #self.optimizer.step()
        #print("before optimizer step")
        #xm.optimizer_step(self.optimizer)
        #xm.mark_step()

        xm.optimizer_step(self.optimizer)

        #tracker.add(8)
	
        #if self.iteration % 5 == 0:
        #    print("++++++ general log ++++++")
        #    xm.add_step_closure(test_utils.print_training_update(device, self.iteration, loss.item(), tracker.rate(),
        #                           tracker.global_rate()), args=(device, self.iteration, loss, tracker))
        #    print("++++++ end of general log ++++++")

        #print("before return")
        return loss#.detach()#.item()

Thank you in advance :)

**Comment by dlibenzi** [comment]
Can you try just one core (pass nprocs=1 to `xmp.spawn()`)?

**Comment by KonstantinosMarko** [comment]
hmm i have never used xmp.spawn()...

is it essential to use it in my code?

**Comment by dlibenzi** [comment]
Do you use DataParallel?
Can you should the outer skeleton of you code?

**Comment by KonstantinosMarko** [comment]
i have 2 versions. one with DataParallel and one with no DataParallel. 
At first, i am trying to find out why in the simplest one (the one with no DataParallel), it freezes after the first epoch.
